src/Bicep.Core/TypeSystem/TypeAssignmentVisitor.cs

```diff
@@ -534,12 +534,22 @@ public override void VisitUnionTypeSyntax(UnionTypeSyntax syntax)
             => AssignTypeWithDiagnostics(syntax, diagnostics =>
             {
                 var declaredType = typeManager.GetDeclaredType(syntax) ?? ErrorType.Empty();
-                diagnostics.WriteMultiple(declaredType.GetDiagnostics());
+
+                if (declaredType is not ErrorType)
+                {
+                    diagnostics.WriteMultiple(declaredType.GetDiagnostics());
+                }
```

Comment:
You can skip this; only `ErrorType`s will ever have diagnostics, and `diagnostics.WriteMultiple(declaredType.GetDiagnostics());` (without checking the type of `declaredType`) will be called when the parent syntax node gets processed in this class.

Reply:
Ok I was uncertain about this. I added this because I was running into duplicate diagnostics, but now I'm wondering if it will cause issues to go unreported. Nothing failed when I made this change, but will retest. I guess it depends on where union type syntax is allowed, which to my current knowledge, is user defined type scenarios, param types, and output types.

Reply:
Having some diagnostics reported by the parent like this isn't ideal, but it's forced on us at the moment by the way declared symbols check for definitional errors. Every symbol declaration will surface diagnostics from its declared type, so the only way I found to avoid duplicate diagnostics was to have each "TypeSyntax parent" that is itself part of a type expression report the diagnostics of its direct child.

So these expressions would have errors reported by the declared symbol error checker, where `X` represents an invalid type expression:

```bicep
type foo = X
param bar X
output baz X
```

Whereas these expressions would have errors reported by their "TypeSyntax parent":

```bicep
type foo = {
  bar: X
  *: X
}
type fizz = [
  X
]
type snap = X[]
```

Reply:
I restored the write diagnostics call and added a `Distinct()` call to the `SemanticModel::GetAllDiagnostics` method which solved the issue of duplicate diagnostics.